### ISSUE TITLE
Disabling session management per request does not work

### ISSUE THREAD

**Issue by andybellenie** [cfwheels/cfwheels#493]
Session management can be disabled for some requests but not others, by putting logic into /config/app.cfm

This is useful to reduce memory footprint if a request doesn't require a session. Think api, heavy workloads etc.

However, wheels stores an application setting for sessionManagement and refers to it in the code. This shouldn't really be an application setting, Application.cfc is instantiated on each request so this.sessionManagement is more accurately request level, not application.

There's an undocumented function (GetApplicationSettings()) which returns the correct values for the current request.

I'll have a go at fixing this. Be good to get my hands dirty again :)

Any objections?

**Comment by perdjurner** [comment]
"wheels stores an application setting for sessionManagement and refers to it in the code"

Can you give me an example of that?

Just so I can understand this better.

Thanks.

**Comment by andybellenie** [comment]
Yes, line 211 of /wheels/events/onapplicationstart.cfm

```
// if session management is enabled in the application we default to storing flash data in the session scope, if not we use a cookie
if (StructKeyExists(this, "sessionManagement") && this.sessionManagement)
	{
		application.$wheels.sessionManagement = true;
		application.$wheels.flashStorage = "session";
	}
	else
	{
		application.$wheels.sessionManagement = false;
		application.$wheels.flashStorage = "cookie";
	}
```

This is then called in $runVerifications() in /wheels/controller/verifies.cfm

```
// only access the session scope when session management is enabled in the app
if (StructIsEmpty(arguments.sessionScope) && get("sessionManagement"))
{
	arguments.sessionScope = session;
}
```

There's some stuff going on with the flash too as this is also configured to use either cookie or session on an application basis. If sessions are disabled for some requests but not others the flash will error.

I've patched my copy of Wheels to automatically decide where to access the flash depending on the sessionManagement status. This works fine for my purposes although I can foresee an edge case where a developer has specified to use cookie for the flash even if sessions are enabled. I don't know why they'd do that, but it's possible! Do you think we need to cover for that case?

**Comment by perdjurner** [comment]
"I've patched my copy of Wheels to automatically decide where to access the flash depending on the sessionManagement status"

But what happens if you set the Flash on a page with session management on and then try to read the value from the Flash on the next page where session management is off?

**Comment by andybellenie** [comment]
The key wouldn't exist in the cookie based flash so no value would be returned. This would be quite weird behaviour to someone who didn't already know of this issue so I don't think that's a path we should take.

I don't know what the answer is at this point. Open to suggestions!

**Comment by perdjurner** [comment]
Ok, how about this:

- If flashStorage is explicitly set to something we respect that and always read / write from the session (if that's what its set to).

- When no setting exists, we read / write to BOTH session and the cookie. We'll always read / write from the session if it's enabled on the current request. The cookie scope sort of becomes a backup to the session scope then. We would need a timestamp added on all writes so we need which scope to read from.

**Comment by perdjurner** [comment]
I've fixed the part of this that I consider to be a bug (that you can't turn off session management per request).

Leaving this open as an enhancement request though since I think my suggestion above is worth looking into further.

**Comment by perdjurner** [comment]
Closing this as we're using the issue tracker for bugs only from now on.
Please implement your feature suggestion as a plugin or make a pull request.
If you need feedback before starting, please consult the Google Group.

Thanks.

**Comment by adamcameron** [comment]
Hate to resurrect an old issue, but this is def just a bug. Decisions about how sessions are handed are a *request-time* consideration, not an app-initialisation-time thing. This is not subjective.

So the code @ https://github.com/cfwheels/cfwheels/blob/develop/vendor/wheels/events/onapplicationstart.cfc#L334-L341, is just wrong:

```
// If session management is enabled in the application we default to storing Flash data in the session scope, if not we use a cookie.
if (StructKeyExists(this, "sessionManagement") && this.sessionManagement) {
    application.$wheels.sessionManagement = true;
    application.$wheels.flashStorage = "session";
} else {
    application.$wheels.sessionManagement = false;
    application.$wheels.flashStorage = "cookie";
}
```

(And this code is the same now as it was when Andy raised the issue; I'm looking at the same codebase ;-).

That you have an issue with how the flash stuff has been implemented to rely on the above bad implementation is... well... implementation detail of the bug fix. Sometimes bugfixes are non-trivial; and fixing architectural design bugs can be a pain in the butt.

But as to how to deal with handling flashes across sequential requests that have different session management? I don't really think you need to. Just document the behaviour. Realistically, if in consecutive requests I go from a session-managed part of the app to a non-session-managed part of the app; things like flash messaging probably don't apply, as it's not likely to be part of a "scripted" user journey. eg: submitting a session-managed GET-handler to a non-session managed POST-handler... won't be a thing. 

Document to the effect of "for Flash messaging to work out of the box, one will need to stick to the same session-management strategy on both source and destination requests". There. Dealt with.

And the fix for the code is to shift that code to a method call in `Application.cfc`'s pseudo-constructor, which is where logic like this is supposed to go.

**Comment by adamcameron** [comment]
(Oh and obvs switch from `application.$wheels.sessionManagement` and `application.$wheels.flashStorage` to something in the `request` scope): I didn't mean to literally cut and paste the code ;-)

**Comment by zainforbjs** [comment]
@adamcameron We have updated and fixed the bug into our version 3.0, please download the latest version and have your tests.LMK if you find anything. Appreciate your support. Thank you and have a nice day.
